client/transaction-pool/graph/src/validated_pool.rs

```diff
@@ -501,14 +502,53 @@ impl<B: ChainApi> ValidatedPool<B> {
 	pub fn status(&self) -> PoolStatus {
 		self.pool.read().status()
 	}
+
+	/// Notify all watchers that transactions in the block with hash have been finalized
+	pub async fn finalized(&self, block_hash: BlockHash<B>) -> Result<(), B::Error> {
+		debug!(target: "txpool", "Attempting to notify watchers of finalization for {}", block_hash);
+		let last_finalized_hash = *self.last_finalized_hash.lock();
+
+		let mut header = self.api.block_header(BlockId::Hash(block_hash.clone()))?
+			.expect("finalized block should have block header; qed");
+
+		while header.hash() != last_finalized_hash {
```

Comment:
This loop will exit with an error in case there is no path between `header` and `last_finalized_hash`.

We have a helper: `sp_blockchain::tree_route` that will nicely tell you all the blocks on the route between any two of them. The current logic does not handle re-orgs correctly or cases where there is no common block at all.